### PR TITLE
Aggregate data source

### PR DESCRIPTION
**Pull request by TG908**
I reworked the app architecture by moving big parts of our massive view controller 😄 into a hierarchy of data sources in preparation for the new UI

### CODE REVIEW

File: TUM Campus App/CardCells/CafeteriaCollectionViewCell.swift [hunk 1]
```diff
@@ -0,0 +1,18 @@
+//
+//  CafeteriaCollectionViewCell.swift
+//  Campus
+//
+//  Created by Tim Gymnich on 03.05.18.
+//  Copyright © 2018 LS1 TUM. All rights reserved.
```
Comment (kordianbruck) [inline TUM Campus App/CardCells/CafeteriaCollectionViewCell.swift]: Do we need this copy right? What is LS1? Why is this code copyrighted to Prof. Brügge?

File: TUM Campus App/CardViewController.swift [hunk 2]
```diff
@@ -113,12 +103,36 @@ extension CardViewController {
         self.navigationItem.titleView = view
     }
     
-    func setupTableView() {
-        refresh.addTarget(self, action: #selector(CardViewController.refresh(_:)), for: UIControlEvents.valueChanged)
-        tableView.addSubview(refresh)
+    @objc func refresh(sender: AnyObject?) {
+        composedDataSource?.refresh()
+//        TODO fix this
```
Comment (kordianbruck) [inline TUM Campus App/CardViewController.swift]: Would be great, to not commit commented code / todos.

File: TUM Campus App/CardCells/CafeteriaCollectionViewCell.swift (hunk 1 above)
Comment (TG908) [reply TUM Campus App/CardCells/CafeteriaCollectionViewCell.swift]: Xcode does this... Should I just replace this with TCA ???

File: TUM Campus App/CardViewController.swift (hunk 2 above)
Comment (TG908) [reply TUM Campus App/CardViewController.swift]: Its a work in progress :D

File: TUM Campus App/CardCells/CafeteriaCollectionViewCell.swift (hunk 1 above)
Comment (kordianbruck) [reply TUM Campus App/CardCells/CafeteriaCollectionViewCell.swift]: No, please just use the GPLv3: https://github.com/TCA-Team/iOS/blob/master/LICENSE

Short part is this for example: https://gist.github.com/ilg-ul/aa6de96c993719667f07

File: TUM Campus App/MVGManager.swift [hunk 3]
```diff
@@ -23,16 +29,36 @@ final class MVGManager: Manager, SimpleSingleManager {
         self.config = config
     }
     
+    func performRequest(maxCache: CacheTime) -> Promise<[Station], APIError> {
+        return config.mvg.doObjectsRequest(
```
Comment (thellmund) [inline TUM Campus App/MVGManager.swift]: doObjectsRequest() is deprecated. Is there any other method to accomplish the desired functionality?

File: TUM Campus App/MVGManager.swift [hunk 4]
```diff
@@ -23,16 +29,36 @@ final class MVGManager: Manager, SimpleSingleManager {
...
             return self.fetch(for: station).map { departures in
                 return DetailedStation(station: station, departures: departures)
             }
         }
     }
     
-    func fetch() -> Response<[Station]> {
+    func fetch() -> Response<[DetailedStation]> {
+        return prefetch().flatMap { stations in
+            return stations.flatMap { station in
```
Comment (thellmund) [inline TUM Campus App/MVGManager.swift]: flatMap() is deprecated. Check if there’s another way to accomplish the desired functionality.

File: TUM Campus App/MVGManager.swift [hunk 5]
```diff
@@ -23,16 +29,36 @@ final class MVGManager: Manager, SimpleSingleManager {
...
+                return self.fetch(for: station).map { departures in
+                    return DetailedStation(station: station, departures: departures)
+                }
+            }.bulk
+        }
+    }
+
```
Comment (thellmund) [inline TUM Campus App/MVGManager.swift]: Remove blank line.

File: TUM Campus App/NewsSpreadCollectionViewCell.swift [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+//  Copyright © 2018 LS1 TUM. All rights reserved.
+//
+
+import UIKit
+
+class NewsSpreadCollectionViewCell: UICollectionViewCell {
+    
+    @IBOutlet weak var imageView: UIImageView!
+    var binding: ImageViewBinding?
+
```
Comment (thellmund) [inline TUM Campus App/NewsSpreadCollectionViewCell.swift]: Remove blank line.

File: TUM Campus App/PersonDetailTableViewController.swift [hunk 7]
```diff
@@ -21,7 +21,7 @@ class PersonDetailTableViewController: UITableViewController, DetailView {
     var addingContact = false
```
Comment (thellmund) [inline TUM Campus App/PersonDetailTableViewController.swift]: Remove blank line.

File: TUM Campus App/PersonDetailTableViewController.swift [hunk 8]
```diff
@@ -21,7 +21,7 @@ class PersonDetailTableViewController: UITableViewController, DetailView {
...
     
     
-    func addContact(_ sender: AnyObject?) {
+    @objc func addContact(_ sender: AnyObject?) {
```
Comment (thellmund) [inline TUM Campus App/PersonDetailTableViewController.swift]: Don’t use AnyObject as argument type. Make explicit or remove if not used.

File: TUM Campus App/RefreshableTableViewController.swift [hunk 9]
```diff
@@ -38,7 +38,7 @@ class RefreshableTableViewController<Value: DataElement>: UITableViewController
         }
     }
     
-    func refresh(_ sender: AnyObject? = nil) {
+    @objc func refresh(_ sender: AnyObject? = nil) {
```
Comment (thellmund) [inline TUM Campus App/RefreshableTableViewController.swift]: Don’t use AnyObject as argument type. Make explicit or remove if not used.

File: TUM Campus App/RoomFinderViewController.swift [hunk 10]
```diff
@@ -84,14 +84,14 @@ extension RoomFinderViewController {
 
 extension RoomFinderViewController {
     
-    func showMaps(_ send: AnyObject?) {
+    @objc func showMaps(_ send: AnyObject?) {
```
Comment (thellmund) [inline TUM Campus App/RoomFinderViewController.swift]: Don’t use AnyObject as argument type. Make explicit or remove if not used.

File: TUM Campus App/RoomFinderViewController.swift [hunk 11]
```diff
@@ -84,14 +84,14 @@ extension RoomFinderViewController {
...
         pickerView.show()
         barItem?.action = #selector(RoomFinderViewController.hideMaps(_:))
         barItem?.image = UIImage(named: "collapse")
         navigationController?.interactivePopGestureRecognizer?.isEnabled = false
     }
     
-    func hideMaps(_ send: AnyObject?) {
+    @objc func hideMaps(_ send: AnyObject?) {
```
Comment (thellmund) [inline TUM Campus App/RoomFinderViewController.swift]: Don’t use AnyObject as argument type. Make explicit or remove if not used.

File: TUM Campus App/StudyRoomsTableViewController.swift [hunk 12]
```diff
@@ -35,7 +35,7 @@ class StudyRoomsTableViewController: UITableViewController, DetailView {
         }
     }
     
-    func refresh(_ sender: AnyObject?) {
+    @objc func refresh(_ sender: AnyObject?) {
```
Comment (thellmund) [inline TUM Campus App/StudyRoomsTableViewController.swift]: Don’t use AnyObject as argument type. Make explicit or remove if not used.

File: TUM Campus App/StudyRoomsTableViewController.swift [hunk 13]
```diff
@@ -102,14 +102,14 @@ extension StudyRoomsTableViewController: DetailViewDelegate {
 
 extension StudyRoomsTableViewController {
     
-    func showRooms(_ send: AnyObject?) {
+    @objc func showRooms(_ send: AnyObject?) {
```
Comment (thellmund) [inline TUM Campus App/StudyRoomsTableViewController.swift]: Don’t use AnyObject as argument type. Make explicit or remove if not used.

File: TUM Campus App/StudyRoomsTableViewController.swift [hunk 14]
```diff
@@ -102,14 +102,14 @@ extension StudyRoomsTableViewController: DetailViewDelegate {
...
         pickerView.show()
         barItem?.action = #selector(StudyRoomsTableViewController.hideRooms(_:))
         barItem?.image = UIImage(named: "collapse")
         navigationController?.interactivePopGestureRecognizer?.isEnabled = false
     }
     
-    func hideRooms(_ send: AnyObject?) {
+    @objc func hideRooms(_ send: AnyObject?) {
```
Comment (thellmund) [inline TUM Campus App/StudyRoomsTableViewController.swift]: Don’t use AnyObject as argument type. Make explicit or remove if not used.

File: TUM Campus App/TuitionTableViewController.swift [hunk 15]
```diff
@@ -88,14 +88,14 @@ extension TuitionTableViewController {
         navigationItem.rightBarButtonItem = barItem
     }
     
-    func showSemesters(_ send: AnyObject?) {
+    @objc func showSemesters(_ send: AnyObject?) {
```
Comment (thellmund) [inline TUM Campus App/TuitionTableViewController.swift]: Don’t use AnyObject as argument type. Make explicit or remove if not used.

File: TUM Campus App/TuitionTableViewController.swift [hunk 16]
```diff
@@ -88,14 +88,14 @@ extension TuitionTableViewController {
...
         pickerView.show()
         barItem?.action = #selector(TuitionTableViewController.hideSemesters(_:))
         barItem?.image = UIImage(named: "collapse")
         navigationController?.interactivePopGestureRecognizer?.isEnabled = false
     }
     
-    func hideSemesters(_ send: AnyObject?) {
+    @objc func hideSemesters(_ send: AnyObject?) {
```
Comment (thellmund) [inline TUM Campus App/TuitionTableViewController.swift]: Don’t use AnyObject as argument type. Make explicit or remove if not used.

**Comment by thellmund** [comment]
Looking very good overall! Some small possibilities for improvement (and some tiny formatting things). 

There are a couple of things to improve from a UI perspective, but let’s tackle them in separate pull requests.

One important note: The project only builds for me if I update the Podfile with 
`pod 'Sweeft', '~> 0.15'`
@TG908 Please update the Podfile accordingly.

File: TUM Campus App/DataSources/CalendarDataSource.swift [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+//  Copyright © 2018 LS1 TUM. All rights reserved.
+//
+
+import UIKit
+
+
+class CalendarDataSource: NSObject, TUMDataSource {
+    
+    var manager: CalendarManager
+    let cellType: AnyClass = CalendarCollectionViewCell.self
```
[the remark this answers is not in the document]
Comment (TG908) [reply TUM Campus App/DataSources/CalendarDataSource.swift]: I am sticking to the apple api which looks like this
```
collectionView.register(cellClass: AnyClass, forCellWithReuseIdentifier: String)
```

File: TUM Campus App/DataSources/ComposedDataSource.swift [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+        delegate?.didBeginRefreshingDataSources()
+        let group = DispatchGroup()
+        dataSources.forEach{$0.refresh(group: group)}
+        group.notify(queue: .main) {
+            self.delegate?.didRefreshDataSources()
+        }
+    }
+    
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+//        TODO
```
[the remark this answers is not in the document]
Comment (TG908) [reply TUM Campus App/DataSources/ComposedDataSource.swift]: Line 77 includes the code that needs to be implemented...

File: TUM Campus App/DataSources/GradesDataSource.swift [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+    
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return data.count
+    }
+    
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: cellReuseID, for: indexPath) as! GradesCollectionViewCell
+        let grade = data[indexPath.row]
+        
+        cell.gradeLabel.text = grade.result
```
[the remark this answers is not in the document]
Comment (TG908) [reply TUM Campus App/DataSources/GradesDataSource.swift]: Much of the formatting is done in the DataSource, to save resources. Also the view should not know about the model.

File: TUM Campus App/DataSources/GradesDataSource.swift [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+        cell.nameLabel.text = grade.name
+        cell.semesterLabel.text = grade.semester
+        
+        switch Double(grade.result.replacingOccurrences(of: ",", with: "."))  {
```
[the remark this answers is not in the document]
Comment (TG908) [reply TUM Campus App/DataSources/GradesDataSource.swift]: I think its better to leave displaying data to the view and interpreting the data to the dataSource

File: TUM Campus App/MVGManager.swift (hunk 3 above)
Comment (TG908) [reply TUM Campus App/MVGManager.swift]: Thats an issue with @nerdsupremacist sweeft library and should be addressed later.

File: TUM Campus App/DataSources/CalendarDataSource.swift (hunk 17 above)
[the remark this answers is not in the document]
Comment (thellmund) [reply TUM Campus App/DataSources/CalendarDataSource.swift]: Interesting. Alright then 👍

File: TUM Campus App/MVGManager.swift (hunk 3 above)
Comment (thellmund) [reply TUM Campus App/MVGManager.swift]: You’re right. For some reason, it looked to me like it was our own method

File: TUM Campus App/DataSources/GradesDataSource.swift (hunk 20 above)
[the remark this answers is not in the document]
Comment (thellmund) [reply TUM Campus App/DataSources/GradesDataSource.swift]: To me, color can be regarded as an attribute of a grade

File: TUM Campus App/DataSources/GradesDataSource.swift (hunk 19 above)
[the remark this answers is not in the document]
Comment (thellmund) [reply TUM Campus App/DataSources/GradesDataSource.swift]: The cell should not include business logic, but I don’t see a problem with passing a data object to a cell

**Comment by TG908** [comment]
> New taxonomy of data sources looking good overall! Some small possibilities for improvement (and some tiny formatting things).
> 
> The updated cards however don’t seem to work as intended.
> 
> Newsspread does not show any news
> Most elements don’t react to clicks (e.g. the movie posters)
> Some cards show information that’s not relevant in the moment (e.g. the TU film in 2 months)
> One important note: The project only builds for me if I update the Podfile with
> pod 'Sweeft', '~> 0.15'
> Please update the Podfile accordingly.


This doesn't make sense. Version 0.15 of Sweeft still has the parts of the code that don't work anymore in > Swift 4.

**Comment by thellmund** [comment]
I’ve made some adjustments to the UI and code formatting. Most importantly, I’ve added the image loading library Kingfisher, as our own implementation does not appear reliable. 

One thing I’m confused by: When created, `News`objects hold an image url. However, once accessing the object in `CollectionView`’s`didSelectItemAt`, that member variable is `nil` again. @TG908, any idea on why that might be the case and how to fix it?

**Comment by TG908** [comment]
Thing left to do before release:

- [ ] Implement card order

- [ ] Show something on card selections

- [ ] Maybe add network timeout and no connection screen

**Comment by thellmund** [comment]
I just created a pull request into this branch to add Firebase integration. Once that is approved, we need to merge this branch into master and start using proper branching again.

File: Podfile [hunk 21]
```diff
@@ -14,6 +14,8 @@ target 'Campus' do
     pod 'SWXMLHash', '~> 4.6.0'
     pod 'TKSubmitTransition', :git => 'https://github.com/entotsu/TKSubmitTransition.git', :branch => 'swift4'
     pod 'Kanna', '~> 4.0.0'
+	pod 'Kingfisher', '~> 4.0'
```
Comment (pfent) [inline Podfile]: mixing tabs and spaces for indentation :tada:

**Pull request merged**